u_tls_extensions.go

```diff
@@ -248,6 +248,7 @@ type ALPNExtension struct {
 func (e *ALPNExtension) writeToUConn(uc *UConn) error {
 	uc.config.NextProtos = e.AlpnProtocols
 	uc.HandshakeState.Hello.AlpnProtocols = e.AlpnProtocols
+	uc.HandshakeState.Hello.NextProtoNeg = len(e.AlpnProtocols) > 0
```

Comment:
Yeah I'm not in love with how this has to happen, but go`s `clientHelloMsg` has a boolean `nextProtoNeg` field. [ALPN](https://tools.ietf.org/html/rfc7301) is an extension for negotiating an application layer protocol. The way that this is normally determined is by way of the underlying `UConn.config`:
https://github.com/getlantern/utls/blob/master/handshake_client.go#L77
```
		nextProtoNeg:                 len(config.NextProtos) > 0,
		alpnProtocols:                config.NextProtos,
```

However, we can't really modify the `config` when just generating the spec. In fact I find this whole logic flow/dependency really strange. I was struggling a bit to understand the interaction between ALPN and the `nextprotoneg` [extension](https://tools.ietf.org/html/draft-agl-tls-nextprotoneg-04). I think maybe this should either be removed or replaced with:
```
uc.HandshakeState.Hello.NextProtoNeg = uc.HandshakeState.Hello.NextProtoNeg || len(e.AlpnProtocols) > 0
```

If I remove it, it fails the tests that check against the `HelloRandomizedALPN` client ids. It's possible that's actually due to a misimplementation in the `generateRandomizedSpec` code. I can look into it a little more...

Reply:
I think this controls the next protocol negotiation extension, not alpn ?https://tools.ietf.org/html/draft-agl-tls-nextprotoneg-04

Reply:
That's interesting because it doesn't seem like you've made any changes which should affect the ALPN logic.  I guess I must be missing something.  Possibly something to flag and ask Sergey about when we try to get this merged upstream.